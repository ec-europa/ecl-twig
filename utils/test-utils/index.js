const Twig = require('twig');
const { format } = require('prettier');

const renderTwigFile = (file, options, cb) =>
  Twig.renderFile(file, options, (err, html) =>
    cb(err, format(html, { parser: 'html' }))
  );

const renderTwigFileAsNode = (file, options) =>
  new Promise((resolve, reject) =>
    Twig.renderFile(file, options, (err, html) => {
      if (err) return reject(err);

      const div = document.createElement('div');
      div.innerHTML = html.trim();

      return resolve(div.firstChild);
    })
  );

module.exports = {
  renderTwigFile,
  renderTwigFileAsNode,
};
