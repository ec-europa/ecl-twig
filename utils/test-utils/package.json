{
  "name": "@ecl-twig/test-utils",
  "version": "2.29.0",
  "description": "A local test helper",
  "main": "index.js",
  "private": true,
  "devDependencies": {
    "deepmerge": "4.2.2"
  }
}
