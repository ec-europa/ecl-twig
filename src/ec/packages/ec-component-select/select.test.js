import path from 'path';
import { merge, renderTwigFileAsNode } from '@ecl-twig/test-utils';

import specData from '@ecl/ec-specs-select/demo/data';

describe('EC - Select', () => {
  const template = path.resolve(__dirname, './select.html.twig');
  const render = params => renderTwigFileAsNode(template, params);

  describe('Default', () => {
    test('renders correctly', () => {
      expect.assertions(1);

      return expect(
        render({
          label: specData.label,
          options: specData.options,
          helper_text: 'Help message',
          icon_path: 'path-to-icons-file.svg',
          id: 'example-id',
          name: 'example-name',
        })
      ).resolves.toMatchSnapshot();
    });
  });

  describe('Disabled', () => {
    test('renders correctly', () => {
      expect.assertions(1);

      return expect(
        render({
          disabled: true,
          label: specData.label,
          options: specData.options,
          helper_text: 'Help message',
          icon_path: 'path-to-icons-file.svg',
          id: 'example-id',
          name: 'example-name',
        })
      ).resolves.toMatchSnapshot();
    });
  });

  describe('With error', () => {
    const options = {
      invalid: true,
      invalid_text: 'Error message',
      label: specData.label,
      options: specData.options,
      helper_text: 'Help message',
      icon_path: 'path-to-icons-file.svg',
      id: 'example-id',
      name: 'example-name',
    };

    test('renders correctly', () => {
      expect.assertions(1);

      return expect(render(options)).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra class names', () => {
      expect.assertions(1);

      const optionsWithExtraClasses = merge(options, {
        extra_classes: 'custom-select custom-select--test',
      });

      return expect(render(optionsWithExtraClasses)).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra attributes', () => {
      expect.assertions(1);

      const optionsWithExtraClasses = merge(options, {
        extra_attributes: [
          { name: 'data-test', value: 'data-test-value' },
          { name: 'data-test-1', value: 'data-test-value-1' },
        ],
      });

      return expect(render(optionsWithExtraClasses)).resolves.toMatchSnapshot();
    });
  });
});
