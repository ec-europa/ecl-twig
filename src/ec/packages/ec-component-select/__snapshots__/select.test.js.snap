// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EC - Select Default renders correctly 1`] = `
<div
  class="ecl-form-group ecl-form-group--select"
>
  <label
    class="ecl-form-label"
    for="example-id"
  >
    Select a country
  </label>
  <div
    class="ecl-select__container"
  >
    <select
      class="ecl-select"
      id="example-id"
      name="example-name"
    >
      <option
        value="1"
      >
        Belgium
      </option>
      <option
        value="2"
      >
        France
      </option>
      <option
        value="3"
      >
        Luxembourg
      </option>
      <option
        value="4"
      >
        Germany
      </option>
      <option
        value="5"
      >
        Bulgaria
      </option>
    </select>
    <div
      class="ecl-select__icon"
    >
      <svg
        aria-hidden="true"
        class="ecl-icon ecl-icon--s ecl-icon--rotate-180 ecl-select__icon-shape"
        focusable="false"
      >
        <use
          xlink:href="path-to-icons-file.svg#ui--corner-arrow"
        />
      </svg>
    </div>
  </div>
  <div
    class="ecl-help-block"
  >
    Help message
  </div>
</div>
`;

exports[`EC - Select Disabled renders correctly 1`] = `
<div
  class="ecl-form-group ecl-form-group--select"
>
  <label
    class="ecl-form-label"
    for="example-id"
  >
    Select a country
  </label>
  <div
    class="ecl-select__container ecl-select__container--disabled"
  >
    <select
      class="ecl-select"
      disabled=""
      id="example-id"
      name="example-name"
    >
      <option
        value="1"
      >
        Belgium
      </option>
      <option
        value="2"
      >
        France
      </option>
      <option
        value="3"
      >
        Luxembourg
      </option>
      <option
        value="4"
      >
        Germany
      </option>
      <option
        value="5"
      >
        Bulgaria
      </option>
    </select>
    <div
      class="ecl-select__icon"
    >
      <svg
        aria-hidden="true"
        class="ecl-icon ecl-icon--s ecl-icon--rotate-180 ecl-select__icon-shape"
        focusable="false"
      >
        <use
          xlink:href="path-to-icons-file.svg#ui--corner-arrow"
        />
      </svg>
    </div>
  </div>
  <div
    class="ecl-help-block"
  >
    Help message
  </div>
</div>
`;

exports[`EC - Select With error renders correctly 1`] = `
<div
  class="ecl-form-group ecl-form-group--select"
>
  <label
    class="ecl-form-label ecl-form-label--invalid"
    for="example-id"
  >
    Select a country
  </label>
  <div
    class="ecl-select__container"
  >
    <select
      class="ecl-select ecl-select--invalid"
      id="example-id"
      name="example-name"
    >
      <option
        value="1"
      >
        Belgium
      </option>
      <option
        value="2"
      >
        France
      </option>
      <option
        value="3"
      >
        Luxembourg
      </option>
      <option
        value="4"
      >
        Germany
      </option>
      <option
        value="5"
      >
        Bulgaria
      </option>
    </select>
    <div
      class="ecl-select__icon"
    >
      <svg
        aria-hidden="true"
        class="ecl-icon ecl-icon--s ecl-icon--rotate-180 ecl-select__icon-shape"
        focusable="false"
      >
        <use
          xlink:href="path-to-icons-file.svg#ui--corner-arrow"
        />
      </svg>
    </div>
  </div>
  <div
    class="ecl-feedback-message"
  >
    Error message
  </div>
  <div
    class="ecl-help-block"
  >
    Help message
  </div>
</div>
`;

exports[`EC - Select With error renders correctly with extra attributes 1`] = `
<div
  class="ecl-form-group ecl-form-group--select"
>
  <label
    class="ecl-form-label ecl-form-label--invalid"
    for="example-id"
  >
    Select a country
  </label>
  <div
    class="ecl-select__container"
  >
    <select
      class="ecl-select ecl-select--invalid"
      data-test="data-test-value"
      data-test-1="data-test-value-1"
      id="example-id"
      name="example-name"
    >
      <option
        value="1"
      >
        Belgium
      </option>
      <option
        value="2"
      >
        France
      </option>
      <option
        value="3"
      >
        Luxembourg
      </option>
      <option
        value="4"
      >
        Germany
      </option>
      <option
        value="5"
      >
        Bulgaria
      </option>
    </select>
    <div
      class="ecl-select__icon"
    >
      <svg
        aria-hidden="true"
        class="ecl-icon ecl-icon--s ecl-icon--rotate-180 ecl-select__icon-shape"
        focusable="false"
      >
        <use
          xlink:href="path-to-icons-file.svg#ui--corner-arrow"
        />
      </svg>
    </div>
  </div>
  <div
    class="ecl-feedback-message"
  >
    Error message
  </div>
  <div
    class="ecl-help-block"
  >
    Help message
  </div>
</div>
`;

exports[`EC - Select With error renders correctly with extra class names 1`] = `
<div
  class="ecl-form-group ecl-form-group--select"
>
  <label
    class="ecl-form-label ecl-form-label--invalid"
    for="example-id"
  >
    Select a country
  </label>
  <div
    class="ecl-select__container"
  >
    <select
      class="ecl-select ecl-select--invalid custom-select custom-select--test"
      id="example-id"
      name="example-name"
    >
      <option
        value="1"
      >
        Belgium
      </option>
      <option
        value="2"
      >
        France
      </option>
      <option
        value="3"
      >
        Luxembourg
      </option>
      <option
        value="4"
      >
        Germany
      </option>
      <option
        value="5"
      >
        Bulgaria
      </option>
    </select>
    <div
      class="ecl-select__icon"
    >
      <svg
        aria-hidden="true"
        class="ecl-icon ecl-icon--s ecl-icon--rotate-180 ecl-select__icon-shape"
        focusable="false"
      >
        <use
          xlink:href="path-to-icons-file.svg#ui--corner-arrow"
        />
      </svg>
    </div>
  </div>
  <div
    class="ecl-feedback-message"
  >
    Error message
  </div>
  <div
    class="ecl-help-block"
  >
    Help message
  </div>
</div>
`;
